### PR TITLE
feat(modal): a11y enhancements (focus order & trapping)

### PR DESCRIPTION
**Pull request by sandgraham**
#### Description

This PR adds a few ARIA a11y enhancements to the Modal component as suggested in issue #1236. 

**Focus trapping**: This means as a user tabs through the Modal, focus will never leave the Modal.

**Focus first interactive element**: When `autofocus` is enabled, the Modal will focus the first focusable element in the Modal. If no interactive elements are available or `autofocus={false}`, the Modal container will be focused.

tabbable

**New `FocusOnce` utility**: A small wrapper component that allows a developer to make an element focus-able for just one focus event. This is useful for the following scenario:
> If content is large enough that focusing the first interactive element could cause the beginning of content to scroll out of view, it is advisable to add tabindex=-1 to a static element at the top of the dialog, such as the dialog title or first paragraph, and initially focus that element.

*[From WAI-ARIA Authoring Practices 1.1](https://www.w3.org/TR/wai-aria-practices/#dialog_modal)*

**Modal close button**: I moved the Modal's close button to be last in the tab order, otherwise it would always be selected first. I also added a focus state to the button so that there would be some visual indication of focus.

Screen Shot 2019-06-14 at 1 26 32 PM

#### Scope

- [ ] Patch: Bug Fix
- [x] Minor: New Feature
- [ ] Major: Breaking Change

#### Todo

- [x] Add documentation for `FocusOnce`
- [x] Update documentation for `Modal` a11y concerns
- [x] Add gifs to PR description

### CODE REVIEW

**Comment by gergelyke** [comment]
works really nice! 👏

File: src/table/filter.js [hunk 1]
```diff
@@ -75,31 +75,33 @@ export default function Filter(props: FilterProps) {
...
-              onClick={() => {
-                onReset();
-              }}
-            >
-              Reset
-            </Button>
-          </Footer>
-        </TrapFocus>
+        <FocusTrap>
+          <div>
```
Comment (nadiia) [inline src/table/filter.js]: What is this `div` added? Any element added has to be not overridable.

Comment (sandgraham) [reply src/table/filter.js]: `FocusTrap` requires a single node be passed and cannot be a Fragment so I wrapped this in a div. The older implementation (`TrapFocus`) would wrapper whatever was inside it in an extra div so I figured it was equivalent here.

**Comment by sandgraham** [comment]
@gergelyke I will need screener approval - the modal now focuses the first interactive element so there are a few visual diffs there.

**Comment by gergelyke** [comment]
> @gergelyke I will need screener approval when you can- the modal now focuses the first interactive element so there are a few visual diffs there.

done

**Comment by sandgraham** [comment]
@chasestarr I don't feel strongly about `Modal` owning `FocusOnce`. I thought about pulling it out, but didn't have another use case in mind.

---

So the a11y recommendation here is to focus the first element in the modal if nothing below it is focusable or if the first interactive element will cause a scroll/jump on focus.

I suppose we could build that in but it adds some ugliness to the Modal container logic for what seems like a minor edge case. I figured giving folks an easy way to address these cases with `FocusOnce` was a nice compromise. That said, I also understand having Modal provide all this by default is a nice enforcement of the a11y recommendations.

I am leaning towards keeping it separate for now but not 10/10 on this.

**Pull request merged**